File: tags/multi2sim-3.0.1/tools/inifile/inifile.py
#!/usr/bin/python

# Copyright (C) 2009 Rafael Ubal Tena
# 
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

import sys
import re


# Global variables

ProgramName = "IniFile 1.0"
FileName = ""  # Name of the file
Modif = False  # Set to true if there was some change
SectionList = []  # Elements = [ SimpleSectionName, SectionName, KeyList ]


def Warning(msg):
	sys.stderr.write("Warning: %s\n" % msg)


def error(msg):
	sys.stderr.write("Error: %s\n" % msg)
	sys.exit(1)


def ReadFile():

	# Flush the list [ SimpleSectionName, SectionName, KeyList ]
	# into SectionList. Those unformatted lines at the end of KeyList
	# are added as sections.
	def FlushSection():
		section_index = len(SectionList)
		while len(KeyList) > 0:
			Key = KeyList[-1]
			[ SimpleKeyName, KeyName, Value ] = Key
			if SimpleKeyName != "":
				break
			Section = [ "", KeyName, [] ]
			SectionList.insert(section_index, Section)
			del(KeyList[-1])
		Section = [ SimpleSectionName, SectionName, KeyList ]
		SectionList.insert(section_index, Section)
	
	# Open file
	try:
		f = open(FileName, "r")
	except:
		return
	
	# Initialize
	global SectionList
	global SectionRe
	global KeyRe
	SimpleSectionName = ""
	SectionName = ""
	KeyList = []
	
	# Read
	for Line in f:
		
		# Remove final '\n'
		if Line != "" and Line[-1] == "\n":
			Line = Line[:-1]
		
		# New section
		SectionMatch = SectionRe.match(Line)
		if SectionMatch:
			if SectionName != "":
				FlushSection()
			SimpleSectionName = SectionMatch.group(1).strip().lower()
			SectionName = Line
			KeyList = []
			continue

		# New key
		KeyMatch = KeyRe.match(Line)
		if KeyMatch:
			SimpleKeyName = KeyMatch.group(1).strip().lower()
			KeyName = Line
			Value = KeyMatch.group(2).strip()
			Key = [ SimpleKeyName, KeyName, Value ]
			KeyList.append(Key)
			continue

		# Comment or unformatted line
		if SectionName == "":
			Section = [ "", Line, [] ]
			SectionList.append(Section)
		else:
			Key = [ "", Line, "" ]
			KeyList.append(Key)
	
	# Store last section and close file
	if SectionName != "":
		FlushSection()
	f.close()


def WriteFile():
	
	# Open file
	try:
		f = open(FileName, "w")
	except:
		error("file '%s' cannot be written" % (FileName))
	
	# Write
	for Section in SectionList:
		[ SimpleSectionName, SectionName, KeyList ] = Section
		f.write("%s\n" % (SectionName))
		for Key in KeyList:
			[ SimpleKeyName, KeyName, Value ] = Key
			f.write("%s\n" % (KeyName))
	
	# Close file
	f.close()


def DumpSection(Section):
	[ SimpleSectionName, SectionName, KeyList ] = Section
	if SimpleSectionName == "":
		print ">>> %s" % SectionName
	else:
		print "Section '%s' ('%s'):" % \
			(SectionName, SimpleSectionName)
		for Key in KeyList:
			[ SimpleKeyName, KeyName, Value ] = Key
			if SimpleKeyName == "":
				print "   >>> %s" % KeyName
			else:
				print "   Key '%s' ('%s' = '%s')" % \
					(KeyName, SimpleKeyName, Value)


# Return the index within SectionList of a given section. If it is not
# present, return -1
def FindSection(Name):
	Name = Name.strip().lower()
	index = 0
	for Section in SectionList:
		[ SimpleSectionName, SectionName, KeyList ] = Section
		if SimpleSectionName == Name:
			return index
		index = index + 1
	return -1


# Return the index within KeyList. If not present, return -1
def FindKey(KeyList, Name):
	Name = Name.strip().lower()
	index = 0
	for Key in KeyList:
		[ SimpleKeyName, KeyName, Value ] = Key
		if SimpleKeyName == Name:
			return index
		index = index + 1
	return -1


# Create a new section at the end of the list and return its index.
def CreateSection():

	# Add a blank line if necessary
	if len(SectionList) > 0:
		Section = SectionList[-1]
		[ SimpleSectionName, SectionName, KeyList ] = Section
		SectionName = SectionName.strip()
		if SectionName != "":
			Section = [ "", "", [] ]
			SectionList.append(Section)
	
	# Add section
	Section = [ "", "", [] ]
	SectionList.append(Section)
	return len(SectionList) - 1


# Create a new key in section with index 'section_index' and return its index.
# The key will be created just after the last present key. */
def CreateKey(section_index):
	
	# Find position
	Section = SectionList[section_index]
	[ SimpleSectionName, SectionName, KeyList ] = Section
	index = -1
	for index in range(len(KeyList) - 1, -1, -1):
		Key = KeyList[index]
		[ SimpleKeyName, KeyName, Value ] = Key
		if SimpleKeyName != "":
			index = index + 1
			break
	index = max(index, 0)

	# Insert and return index
	Key = [ "", "", "" ]
	KeyList.insert(index, Key)
	Section = [ SimpleSectionName, SectionName, KeyList ]
	SectionList[section_index] = Section
	return index


def ProcessCommandDump(Args):
	if len(Args) == 0:
		for Section in SectionList:
			DumpSection(Section)
	
	elif len(Args) == 1:
		index = FindSection(Args[0])
		if index < 0:
			error("section '%s' not found" % Args[0])
		DumpSection(SectionList[index])
	else:
		error("command 'dump': wrong syntax")


def ProcessCommandRead(Args):
	
	# Syntax
	if len(Args) != 2 and len(Args) != 3:
		error("command 'read': wrong syntax")
	
	# Default return value
	Def = ""
	if len(Args) == 3:
		Def = Args[2]
	
	# Find section
	index = FindSection(Args[0])
	if index < 0:
		print Def
		return
	Section = SectionList[index]
	[ SimpleSectionName, SectionName, KeyList ] = Section

	# Find key
	index = FindKey(KeyList, Args[1])
	if index < 0:
		print Def
		return
	Key = KeyList[index]
	[ SimpleKeyName, KeyName, Value ] = Key
	print Value


def ProcessCommandWrite(Args):
	
	# Syntax
	global Modif
	if len(Args) != 3:
		error("command 'write': wrong syntax")
	Args[0] = Args[0].strip()
	Args[1] = Args[1].strip()
	Args[2] = Args[2].strip()
	if len(Args[0]) == 0:
		error("section name not valid: '%s'" % (Args[0]))
	if len(Args[1]) == 0:
		error("key name not valid: '%s'" % (Args[1]))

	# Find section or create new one
	section_index = FindSection(Args[0])
	if section_index < 0:
		section_index = CreateSection()
		SimpleSectionName = Args[0].lower()
		SectionName = "[ " + Args[0] + " ]"
		KeyList = []
		Section = [ SimpleSectionName, SectionName, KeyList ]
		SectionList[section_index] = Section
	else:
		Section = SectionList[section_index]
		[ SimpleSectionName, SectionName, KeyList ] = Section
	
	# Find key or create new one
	key_index = FindKey(KeyList, Args[1])
	if key_index < 0:
		key_index = CreateKey(section_index)
		Section = SectionList[section_index]
		[ SimpleSectionName, SectionName, KeyList ] = Section
	SimpleKeyName = Args[1].lower()
	KeyName = Args[1] + " = " + Args[2]
	Value = Args[2]
	Key = [ SimpleKeyName, KeyName, Value ]
	KeyList[key_index] = Key
	Section = [ SimpleSectionName, SectionName, KeyList ]
	SectionList[section_index] = Section
	
	# File modified
	Modif = True


def ProcessCommandRemove(Args):
	
	# Syntax
	global Modif
	if len(Args) != 1 and len(Args) != 2:
		error("command 'remove': wrong syntax")
	
	# Find section
	section_index = FindSection(Args[0])
	if section_index < 0:
		return
	Section = SectionList[section_index]
	[ SimpleSectionName, SectionName, KeyList ] = Section
	
	# Remove a key
	if len(Args) == 2:
		key_index = FindKey(KeyList, Args[1])
		if key_index < 0:
			return
		del(KeyList[key_index])
		Section = [ SimpleSectionName, SectionName, KeyList ]
		SectionList[section_index] = Section
		Modif = True
		return
	
	# Remove the whole section and following blank lines
	del(SectionList[section_index])
	while section_index < len(SectionList):
		Section = SectionList[section_index]
		[ SimpleSectionName, SectionName, KeyList ] = Section
		SectionName = SectionName.strip()
		if SectionName != "":
			break
		del(SectionList[section_index])
	Modif = True


def ProcessCommandExists(Args):
	
	# Syntax
	if len(Args) != 1 and len(Args) != 2:
		error("command 'exists': wrong syntax")
	Args[0] = Args[0].strip()
	if len(Args) == 2:
		Args[1] = Args[1].strip()
	
	# Find section
	section_index = FindSection(Args[0])
	if section_index < 0:
		print 0
		return
	if len(Args) == 1:
		print 1
		return
	Section = SectionList[section_index]
	[ SimpleSectionName, SectionName, KeyList ] = Section
	
	# Find key
	key_index = FindKey(KeyList, Args[1])
	if key_index < 0:
		print 0
		return
	print 1


def ProcessCommandList(Args):
	
	# Syntax
	if len(Args) != 0 and len(Args) != 1:
		error("command 'list': wrong syntax")
	
	# List sections
	if len(Args) == 0:
		for Section in SectionList:
			[ SimpleSectionName, SectionName, KeyList ] = Section
			if SimpleSectionName != "":
				SectionMatch = SectionRe.match(SectionName)
				print SectionMatch.group(1)
		return
	
	# List keys
	section_index = FindSection(Args[0])
	if section_index < 0:
		return
	Section = SectionList[section_index]
	[ SimpleSectionName, SectionName, KeyList ] = Section
	for Key in KeyList:
		[ SimpleKeyName, KeyName, Value ] = Key
		if SimpleKeyName != "":
			KeyMatch = KeyRe.match(KeyName)
			print KeyMatch.group(1)


def ParseString(s):
	
	def FlushString(temp, end_char):
		temp = temp.strip()
		if len(temp) > 0 or end_char == "\"":
			array.append(temp)
	
	array = []
	temp = ""
	end_char = " "
	while len(s):
		c = s[0]
		s = s[1:]
		if c == end_char:
			FlushString(temp, end_char)
			temp = ""
			end_char = " "
		elif c == "\"":
			end_char = "\""
		else:
			temp = temp + c
	FlushString(temp, end_char)
	return array



def ProcessCommandRun(Args):
	
	# Syntax
	if len(Args) != 1:
		error("command 'run': wrong syntax")
	
	# Open script
	try:
		f = open(Args[0], "r")
	except:
		error("cannot open script '%s'" % (Args[0]))
	
	# Read it
	for Line in f:
		if Line != "" and Line[-1] == "\n":
			Line = Line[:-1]
		Line = Line.strip()
		Args = ParseString(Line)
		if len(Args) > 1:
			ProcessCommand(Args[0], Args[1:])
	f.close()


def ProcessCommand(Command, Args):
	Command = Command.strip().lower()

	if Command == "dump":
		ProcessCommandDump(Args)
	
	elif Command == "read":
		ProcessCommandRead(Args)
	
	elif Command == "write":
		ProcessCommandWrite(Args)
	
	elif Command == "remove":
		ProcessCommandRemove(Args)
	
	elif Command == "exists":
		ProcessCommandExists(Args)
	
	elif Command == "list":
		ProcessCommandList(Args)
	
	elif Command == "run":
		ProcessCommandRun(Args)
	
	else:
		error("invalid command: %s" % Command)



# Main program

Syntax = ProgramName + """
Edit/read configuration files with sections and keys.

Syntax: inifile <file> <cmd> [<args>]
       <file>       The configuration file.
       <cmd>        The command to perform {read|write|exists|list}.
       <args>       Arguments depending on the command.

The following is a list of possible commands with their associated options.
  
  * Command 'dump'. Arguments: [<section>]
       Dump the contents of the file or of a section if <section> is
       given.

  * Command 'read'. Arguments: <section> <key> [<def>]
       Read the key <key> at section <section>. If <def> is given and the
       section or key to read does not exist, <def> is returned as default
       value. If <key> does not exist and <def> is not given, no error is
       shown and an empty output is given.

  * Command 'write'. Arguments: <section> <key> <val>
       Write the pair '<key> = <val>' in section <section>. If the section
       does not exist, it will be created.

  * Command 'remove'. Arguments: <section> [<key>]
       Remove a whole section or just the key within a section if <key>
       is given. If the section/key is not found, no error is given.

  * Command 'exists'. Arguments: <section> [<key>]
       Check if a section (or a key if <key> is given) exists. Return 1
       in case it does, or 0 otherwise.

  * Command 'list'. Arguments: [<section>]
       If no argument is given, list all sections within the file.
       If <section> is given, list all keys within the section.

  * Command 'run'. Arguments: <script>
       Run a sequence of commands from a script.

"""

if len(sys.argv) < 3:
	sys.stderr.write(Syntax)
	sys.exit(1)

SectionRe = re.compile("^ *\[ *(.*) *\] *$")
KeyRe = re.compile("^ *(.*) *= *(.*) *$")

FileName = sys.argv[1]
Command = sys.argv[2]
Args = sys.argv[3:]

ReadFile()
ProcessCommand(Command, Args)
if Modif:
	WriteFile()

File: branches/KeplerEmu/src/lib/cpp/IniFile.cc
/*
 *  Multi2Sim
 *  Copyright (C) 2013  Rafael Ubal (ubal@ece.neu.edu)
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

#include <algorithm>
#include <cassert>
#include <fstream>
#include <sstream>

#include "IniFile.h"
#include "Misc.h"


namespace misc
{


void IniFile::ItemToSectionVar(const std::string &item, std::string &section,
		std::string &var)
{
	std::vector<std::string> tokens;

	// Extract tokens
	StringTokenize(item, tokens, "|");
	assert(tokens.size() == 1 || tokens.size() == 2);
	section = tokens[0];
	var = tokens.size() == 2 ? tokens[1] : "";
}


std::string IniFile::SectionVarToItem(const std::string &_section,
		const std::string &_var)
{
	// Trim section and variable strings
	std::string section = _section;
	std::string var = _var;
	StringSingleSpaces(section);
	StringTrim(var);

	// Check section name
	if (!section.length())
		panic("%s: invalid section name", __FUNCTION__);

	// Create return std::string
	std::string item = section;
	if (var.length())
		item += '|' + var;
	
	// Return
	return item;
}


bool IniFile::GetVarValue(const std::string &s, std::string &var,
		std::string &value)
{
	// At least one '-' sign
	size_t count = std::count(s.begin(), s.end(), '=');
	if (count != 1)
		return false;

	// Split std::string
	size_t pos = s.find('=');
	var = s.substr(0, pos);
	value = s.substr(pos + 1);
	if (var.empty())
		return false;

	// Trim them
	StringTrim(var);
	StringTrim(value);
	return true;
}


/* Insert a new section into the 'item_table' hash table. If the section already
 * exists, return false. Return true on success. */
bool IniFile::InsertSection(std::string section)
{
	bool exists;

	// Get sections
	StringSingleSpaces(section);
	exists = items.count(section);

	// Add section
	items[section] = "";
	if (!exists)
		sections.push_back(section);

	// Return value
	return !exists;
}


/* Insert a variable and its value into the 'items' hash table.
 * If 'var' already existed, the previous value is replaced by the
 * new value, returning false. If 'var' did not exist, return true. */
bool IniFile::InsertVariable(std::string section, std::string var, std::string value)
{
	std::string item;
	bool exists;

	// Combine section and variable
	item = SectionVarToItem(section, var);

	// Set new value
	exists = items.count(item);
	items[item] = value;
	return !exists;
}


IniFile::IniFile()
{
	path = "<IniFile>";
}


IniFile::IniFile(const std::string &path)
{
	Load(path);
}


void IniFile::Dump(std::ostream &os) const
{
	std::string tmp_section;
	std::string section;
	std::string item;
	std::string var;
	std::string value;

	for (auto it = sections.begin(); it != sections.end(); ++it)
	{
		section = *it;
		os << "[ " << section << " ]\n";
		for (auto it2 = items.begin(); it2 != items.end(); ++it2)
		{
			std::string item = it2->first;
			ItemToSectionVar(item, tmp_section, var);
			if (!strcasecmp(section.c_str(), tmp_section.c_str())
					&& item.length() > section.length()
					&& item[section.length()] == '|')
				os << var << " = " << it2->second << '\n';
		}
		os << '\n';
	}
}


static const char *ini_file_err_format =
	"\tA syntax error was detected while parsing a configuration INI file.\n"
	"\tThese files are formed of sections in brackets (e.g. '[ SectionName ]')\n"
	"\tfollowed by pairs 'VariableName = Value'. Comments preceded with ';'\n"
	"\tor '#' characters can be used, as well as blank lines. Please verify\n"
	"\tthe integrity of your input file and retry.\n";


void IniFile::Load(const std::string &path)
{
	std::string line;
	std::string section;
	std::string var;
	std::string value;

	int line_num;
	bool ok;

	// Open file
	this->path = path;
	std::ifstream f(path);
	if (!f)
		fatal("%s: cannot read from file", path.c_str());
	
	// Read lines
	line_num = 0;
	section = "";
	while (getline(f, line))
	{
		// One more line
		line_num++;

		// Comment or blank line
		StringTrim(line);
		if (!line.length() || line[0] == ';' || line[0] == '#')
			continue;
		
		// New "[ <section> ]" entry
		if (line[0] == '[' && line[line.length() - 1] == ']')
		{
			// Get section name
			section = line.substr(1, line.length() - 2);
			StringSingleSpaces(section);

			// Insert section
			ok = InsertSection(section);
			if (!ok)
				fatal("%s: line %d: duplicated section '%s'.\n%s",
					path.c_str(), line_num,
					section.c_str(), ini_file_err_format);

			// Done for this line
			continue;
		}

		// Check that there is an active section
		if (section == "")
			fatal("%s: line %d: section name expected.\n%s",
				path.c_str(), line_num, ini_file_err_format);
		
		// New "<var> = <value>" entry.
		ok = GetVarValue(line, var, value);
		if (!ok)
			fatal("%s: line %d: invalid format.\n%s",
				path.c_str(), line_num, ini_file_err_format);

		// New variable
		ok = InsertVariable(section, var, value);
		if (!ok)
			fatal("%s: line %d: duplicated variable '%s'.\n%s",
				path.c_str(), line_num, var.c_str(),
				ini_file_err_format);
	}
	
	// End
	f.close();
}


void IniFile::Save(const std::string &path) const
{
	// Open file
	std::ofstream f(path);
	if (!f)
		fatal("%s: cannot write to file", path.c_str());
	
	// Dump
	Dump(f);
	f.close();
}


bool IniFile::Exists(const std::string &_section) const
{
	std::string section = _section;
	StringSingleSpaces(section);
	return items.count(section);
}


bool IniFile::Exists(const std::string &section,
		const std::string &var) const
{
	std::string item = SectionVarToItem(section, var);
	return items.count(item);
}


bool IniFile::Remove(const std::string &_section)
{
	// Remove spaces
	std::string section = _section;
	StringSingleSpaces(section);

	// Remove section and variables
	for (auto it = items.begin(); it != items.end(); )
	{
		// Item is section name
		std::string item = it->first;
		if (!strcasecmp(item.c_str(), section.c_str()) ||
				(!strncasecmp(item.c_str(), section.c_str(), section.length())
				&& item[section.length()] == '|'))
		{
			it = items.erase(it);
			continue;
		}

		// Next element
		++it;
	}

	// Remove section from list
	for (auto it = sections.begin(); it != sections.end(); ++it)
	{
		if (!strcasecmp(it->c_str(), section.c_str()))
		{
			sections.erase(it);
			return true;
		}
	}

	// Section not found
	return false;
}


bool IniFile::Remove(const std::string &section, const std::string &var)
{
	std::string item = SectionVarToItem(section, var);
	return items.erase(item);
}


void IniFile::WriteString(const std::string &section, const std::string &var,
		const std::string &value)
{
	// Create item, section, and variable strings
	std::string item = SectionVarToItem(section, var);
	
	// Insert section and variable into table of allowed items
	allowed_items.insert(section);
	allowed_items.insert(item);

	// Write value
	InsertSection(section);
	InsertVariable(section, var, value);
}


void IniFile::WriteInt(const std::string &section, const std::string &var,
		int value)
{
	char buffer[100];
	
	snprintf(buffer, sizeof buffer, "%d", value);
	WriteString(section, var, buffer);
}


void IniFile::WriteInt64(const std::string &section, const std::string &var,
		long long value)
{
	char buffer[100];
	
	snprintf(buffer, sizeof buffer, "%lld", value);
	WriteString(section, var, buffer);
}


void IniFile::WriteBool(const std::string &section, const std::string &var,
		bool value)
{
	char buffer[100];
	
	snprintf(buffer, sizeof buffer, "%s", value ? "True" : "False");
	WriteString(section, var, buffer);
}


void IniFile::WriteDouble(const std::string &section, const std::string &var,
		double value)
{
	char buffer[100];
	
	snprintf(buffer, sizeof buffer, "%f", value);
	WriteString(section, var, buffer);
}


void IniFile::WriteEnum(const std::string &section, const std::string &var,
		int value, StringMap &map)
{
	std::string s;
	bool error;

	// Translate value
	s = map.MapValue(value, error);
	if (error)
		fatal("%s: invalid value for enumeration (%d)",
				__FUNCTION__, value);
	
	// Write
	WriteString(section, var, s);
}


void IniFile::WritePointer(const std::string &section, const std::string &var,
		void *value)
{
	char buffer[100];
	
	snprintf(buffer, sizeof buffer, "%p", value);
	WriteString(section, var, buffer);
}


std::string IniFile::ReadString(const std::string &section,
		const std::string &var, const std::string &def)
{
	Allow(section, var);
	std::string item = SectionVarToItem(section, var);
	auto it = items.find(item);
	return it == items.end() ? def : it->second;
}


int IniFile::ReadInt(const std::string &section, const std::string &var,
		int def)
{
	std::string value;
	StringError error;
	int result;

	// Obtain value
	value = ReadString(section, var);
	if (value == "")
		return def;

	// Interpret
	result = StringToInt(value, error);
	if (error)
		fatal("%s: section [%s], variable '%s', value '%s': %s\n",
				path.c_str(), section.c_str(), var.c_str(),
				value.c_str(), StringErrorToString(error));

	// Return
	return result;
}


long long IniFile::ReadInt64(const std::string &section,
		const std::string &var, long long def)
{
	std::string value;
	StringError error;
	long long result;

	// Obtain value
	value = ReadString(section, var);
	if (value == "")
		return def;

	// Interpret
	result = StringToInt64(value, error);
	if (error)
		fatal("%s: section [%s], variable '%s', value '%s': %s\n",
				path.c_str(), section.c_str(), var.c_str(),
				value.c_str(), StringErrorToString(error));

	// Return
	return result;
}


bool IniFile::ReadBool(const std::string &section, const std::string &var,
		bool def)
{
	std::string s;

	// Read variable
	s = ReadString(section, var);
	if (s == "")
		return def;

	// True
	if (!strcasecmp(s.c_str(), "t") || !strcasecmp(s.c_str(), "True")
			|| !strcasecmp(s.c_str(), "On"))
		return true;
	
	// False
	if (!strcasecmp(s.c_str(), "f") || !strcasecmp(s.c_str(), "False")
			|| !strcasecmp(s.c_str(), "Off"))
		return false;

	// Invalid value
	fatal("%s: section [%s], variable '%s', invalid value '%s'\n"
			"\tPossible values are {t|True|On|f|False|Off}\n",
			path.c_str(), section.c_str(), var.c_str(), s.c_str());
	return false;
}


double IniFile::ReadDouble(const std::string &section,
		const std::string &var, double def)
{
	std::istringstream ss;
	std::string s;
	double value;

	// Read value
	s = ReadString(section, var);
	if (s == "")
		return def;

	// Convert
	ss.str(s);
	ss >> value;
	if (!ss || !ss.eof())
		fatal("%s: section [%s], variable '%s', invalid double value '%s'\n",
				path.c_str(), section.c_str(), var.c_str(), s.c_str());
	
	// Convert
	return value;
}


int IniFile::ReadEnum(const std::string &section, const std::string &var,
		StringMap &map, int def)
{
	std::string s;

	int value;
	bool error;

	// Read value
	s = ReadString(section, var);
	if (s == "")
		return def;
	
	// Convert
	value = map.MapStringCase(s, error);
	if (!error)
		return value;

	// Error, show options
	fatal("%s: section [%s], variable '%s', invalid value '%s'\n"
			"\tPossible values are %s",
			path.c_str(), section.c_str(), var.c_str(), s.c_str(),
			map.toString().c_str());
	return 0;
}


void *IniFile::ReadPointer(const std::string &section, const std::string &var,
		void *def)
{
	std::string s;
	void *value;

	// Read value
	s = ReadString(section, var);
	if (s == "")
		return def;
	
	// Convert
	sscanf(s.c_str(), "%p", &value);
	return value;
}


void IniFile::Allow(const std::string &_section)
{
	std::string section = _section;
	StringSingleSpaces(section);
	allowed_items.insert(section);
}


void IniFile::Allow(const std::string &section, const std::string &var)
{
	Allow(section);
	std::string item = SectionVarToItem(section, var);
	allowed_items.insert(item);
}


void IniFile::Enforce(const std::string &_section)
{
	std::string section = _section;
	StringSingleSpaces(section);
	allowed_items.insert(section);
	enforced_items.insert(section);
}


void IniFile::Enforce(const std::string &section, const std::string &var)
{
	Enforce(section);
	std::string item = SectionVarToItem(section, var);
	enforced_items.insert(item);
	allowed_items.insert(item);
}


void IniFile::Check() const
{
	std::string item;
	std::string section;
	std::string var;

	// Check that all mandatory items are present
	for (auto it = enforced_items.begin(); it != enforced_items.end(); ++it)
	{
		// Item is present
		item = *it;
		if (items.count(item))
			continue;

		// Item not present
		ItemToSectionVar(item, section, var);
		if (var == "")
			fatal("%s: section [%s] missing",
				path.c_str(), section.c_str());
		else
			fatal("%s: section [%s], variable '%s' missing",
				path.c_str(), section.c_str(), var.c_str());
	}

	// Check that all present items are allowed
	for (auto it = items.begin(); it != items.end(); ++it)
	{
		// Item is allowed
		item = it->first;
		if (allowed_items.count(item))
			continue;

		// Item not allowed
		ItemToSectionVar(item, section, var);
		if (var == "")
			fatal("%s: invalid section [%s]",
				path.c_str(), section.c_str());
		else
			fatal("%s: section [%s], invalid variable '%s'",
				path.c_str(), section.c_str(), var.c_str());
	}
}


}  // namespace misc

